Added declaration and initialization of a list of beamline elements.

cemitch99 opened the pull request:
This is a first attempt at initializing a list of beamline elements defining the lattice.  For now, the list includes only drift elements.  The main work is done in the function initElements() in ImpactX.cpp.  

Request sanity check that this makes some sense before proceeding further.

cemitch99 committed:
Added declaration and initialization of a list of beamline elements.

cemitch99 committed:
Removed whitespace from ImpactX.cpp following failed Doxygen style test.

ax3l commented on src/ImpactX.cpp:
```diff
@@ -117,4 +117,18 @@ namespace impactx
 
+    void ImpactX::initElements ()
+    {
+        // initialize element sequence
+        myelements = { };
+
```
The list is empty by default: this constructor is called in `ImpactX::ImpactX` implicitly when the class that owns this member variable is constructed.
Thus, we can skip it here :)
```suggestion
```

ax3l replied:
Ah, you want to make sure they are cleared. That works :)

ax3l commented on src/ImpactX.cpp:
```diff
@@ -117,4 +117,18 @@ namespace impactx
+        myelements = { };
+
+        // add elements
+        Drift drift1(1.0);
+        Drift drift2(0.5);
+        myelements.push_back(drift1);
+        myelements.push_back(drift2);
```
We can even make it more compact and avoid a copy by using:
```suggestion
        myelements.emplace_back(Drift(1.0));
        myelements.emplace_back(Drift(0.5));
```

ax3l commented on src/ImpactX.cpp:
```diff
@@ -117,4 +117,18 @@ namespace impactx
         } // end step loop
     }
 
+    void ImpactX::initElements ()
```
This new function also needs to be declared in `ImpactX.H` :)

ax3l commented on src/ImpactX.H:
```diff
@@ -9,10 +9,14 @@
+
 #include <AMReX_AmrCore.H>
 
 #include <memory>
 
+#include <list>
```
Just grouping and sorting includes:
```suggestion
#include <list>
#include <memory>

```

ax3l commented on src/ImpactX.H:
```diff
@@ -9,10 +9,14 @@
 
 #include "particles/ImpactXParticleContainer.H"
 
+#include "particles/elements/Drift.H"
```
Grouped and sorted:
```suggestion
#include "particles/elements/Drift.H"
#include "particles/ImpactXParticleContainer.H"
```

ax3l commented on src/ImpactX.cpp:
```diff
@@ -117,4 +117,18 @@ namespace impactx
 
+    void ImpactX::initElements ()
+    {
+        // initialize element sequence
+        myelements = { };
```
The list will be automatically constructed in `ImpactX::ImpactX` as an empty container.
Here, it's good to make sure the list is empty, e.g., if we want to re-initialize it.

`= {}` is semantically equivalent to `.clear()`, but the latter might be a bit more explicit:
```suggestion
        // make sure the element sequence is empty
        myelements.clear();
```

ax3l committed:
Drift List: Smaller Cleanup

ax3l committed:
Elements: Use std::variant

This allows us to create a list of multiple (known) types.

ax3l committed:
Push: Loop Over All Elements

ax3l committed:
Workaround: NVCC 11 Nested Device Lambda

CUDA builds for NVCC 11.0.2 fail with
```
  error #3206-D: An extended __device__ lambda cannot be defined inside a generic lambda expression("operator()").
```
if we try to nest out `std::visit` and `ParallelFor` device lambdas.
So, as a work-around, we isolate the single particle push out into a
C++ functor.

Wow, 2010 feels.

ax3l commented:
Work-around for CUDA: submitted as Nvidia request for Enhancement with BugID 3458976

Tiny demonstrator in https://cuda.godbolt.org/z/39e4q53Ye
